### PR TITLE
[WebDriver BiDi] add tests for browsingContext.setViewport

### PR DESCRIPTION
**Pull request by OrKoN**
Proposes spec change: https://github.com/w3c/webdriver-bidi/pull/425

### CODE REVIEW

**Comment by whimboo** [comment]
@OrKoN would you mind adding invalid scenarios as well to `invalid.py` so we have a proper coverage for expected errors as well? Thanks.

**Comment by OrKoN** [comment]
@whimboo I would do it in a separate PR to keep PRs small

File: webdriver/tests/bidi/browsing_context/set_viewport/__init__.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from webdriver.bidi.modules.script import ContextTarget
+
+async def layout_viewport_dimensions(bidi_session, context):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/__init__.py:3]: In `webdriver/tests/bidi/input/perform_actions/__init__.py` we already have a method called `get_viewport_rect`. Maybe we could make that a fixture for all BiDi tests and call it `viewport_dimensions`?

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    await bidi_session.browsing_context.navigate(
+        context=top_context["context"], url="about:blank", wait="complete"
+    )
+
+    await bidi_session.browsing_context.set_viewport(
+        context=top_context["context"],
+        viewport={
+          "width": 250,
+          "height": 300,
+      })
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:20]: So https://pptr.dev/api/puppeteer.page.setviewport suggests to set the viewport dimensions before the navigation. So it means this restriction doesn't actually exist, right and the viewport will immediately be updated and accordingly the layout gets flushed?

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import pytest
+
+from tests.support.image import png_dimensions
+
+from . import layout_viewport_dimensions
+
+@pytest.mark.asyncio
+async def test_capture(bidi_session, top_context, inline):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:8]: Could you please split this test into different scenarios?

* Only setting the width
* Only setting the height
* Setting the width and height
* Resetting the view port to it's original size
* setting the viewport only affects the current tab
* viewport settings persist after a reload / navigation (with and without browsing context replacement - COOP navigation)

For the first 3 tests we could use parameterization.

Also I think that each of the tests should as best run in it's own tab so that it's all cleaned up correctly after the test has been run.

Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:8]: Oh, and maybe you could also add a new test under capture screenshot which tests that as well? Reading the proposed test name triggered such a test for me.

**Review by whimboo — CHANGES_REQUESTED** [review]
Thanks Alex. If you take a look at the other existing tests for other commands we are trying to create as minimal test cases as possible, each covering a different scenario. This makes it easier to verify all the different combinations work as expected, the test coverage is higher in case of something fails early, and each of us can easier mark specific tests as expected fail. I've given some suggestions inline for such a set of tests.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 2 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:18]: I think that refers to websites that read viewport dimensions once upon load and then render things depending on the initial viewport. So although the browser would update everything as expected, the site might not catch up with this without a reload.

File: webdriver/tests/bidi/browsing_context/set_viewport/__init__.py (hunk 1 above)
Comment (thiagowfx) [reply webdriver/tests/bidi/browsing_context/set_viewport/__init__.py:3]: https://sourcegraph.com/github.com/web-platform-tests/wpt/-/blob/webdriver/tests/bidi/browsing_context/capture_screenshot/__init__.py already has one too.

The best course of action here would be to create a fixture (or a helper method put in a more central place) and reuse it everywhere

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 4]
```diff
@@ -127,3 +127,17 @@ def print(self,
...
+    def set_viewport(self, context: str, viewport: Optional[Mapping[str, Any]] = None) -> Mapping[str, Any]:
+        params: MutableMapping[str, Any] = {"context": context}
+
+        if viewport is not None:
+            params["viewport"] = {
+                "width": viewport["width"],
+                "height": viewport["height"]
+            }
+        else:
+            params["viewport"] = None
```
[the remark this answers is not in the document]
Comment (whimboo) [reply tools/webdriver/webdriver/bidi/modules/browsing_context.py:141]: No, we have to set `None` / `null` to actually reset the viewport emulation.

[the remark this answers is not in the document]
Comment (OrKoN) [reply tools/webdriver/webdriver/bidi/modules/browsing_context.py:141]: We could do `params["viewport"] = viewport` but I am not sure what style is preferred here. I think it is better to be more explicit although the code is a bit redundant. WDYT?

File: webdriver/tests/bidi/browsing_context/set_viewport/__init__.py (hunk 1 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/__init__.py:3]: Should I do the refactoring before this PR? capture_screenshot/__init__.py could be split into get_viewport_rect and `get_dpr` and those can be combined in the test.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 3 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py]: I think we do not support setting only the width or only the height, currently?  I wonder if we need to perhaps add it to the spec. The behaviour in Chrome is that zero value means the original size.

File: webdriver/tests/bidi/browsing_context/set_viewport/__init__.py (hunk 1 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/__init__.py:3]: Does anyone know why `get_viewport_rect` falls back to the `document.documentElement`? 

```
    expression = """
        ({
          height: window.innerHeight || document.documentElement.clientHeight,
          width: window.innerWidth || document.documentElement.clientWidth,
        });
    """
```

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thiagowfx) [reply tools/webdriver/webdriver/bidi/modules/browsing_context.py:141]: Then it's OK to leave it as is.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 3 above)
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py]: To open a new tab for each test you can use the `new_tab` fixture. Quite a lot of wdspec tests for BiDi make use of it.

File: webdriver/tests/bidi/browsing_context/set_viewport/__init__.py (hunk 1 above)
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/__init__.py:3]: So this is not needed anymore given that we have the global fixture for bidi now.

**Review by whimboo — CHANGES_REQUESTED** [review]

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 3 above)
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py]: Yes have a look at the following classic test which exercises all the variants. I assume we only need a normal and coop navigation: https://github.com/web-platform-tests/wpt/blob/master/webdriver/tests/classic/navigate_to/navigate.py#L42-L53.

Reload is not implemented in Firefox yet but @thiagowfx is currently working on adding more reload tests. See https://github.com/web-platform-tests/wpt/pull/40134 for example. So I assume that for Chrome it should work? If not we could add a todo item and add it later.

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py]: Reload seems to work with chromium-bidi 👍

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    await bidi_session.browsing_context.set_viewport(
+        context=new_tab["context"],
+        viewport=test_viewport)
+    
+    assert await get_viewport_dimensions(bidi_session, new_tab) == test_viewport
+
+    url = inline("<div>foo</div>")
+    result = await bidi_session.browsing_context.navigate(
+        context=new_tab["context"], url=inline("<div>foo</div>"), wait="complete"
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:58]:
```suggestion
    result = await bidi_session.browsing_context.navigate(
        context=new_tab["context"], url=url, wait="complete"
```

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
+import pytest
+
+from ... import get_viewport_dimensions
+
+@pytest.mark.asyncio
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:5]:
```suggestion

@pytest.mark.asyncio
```

Again please use two empty lines as separator for all the top-level methods including here.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    await bidi_session.browsing_context.set_viewport(
+      context=new_tab["context"],
+      viewport=test_viewport)
+
+    assert await get_viewport_dimensions(bidi_session, new_tab) == test_viewport
+    assert await get_viewport_dimensions(bidi_session, top_context) == original_viewport
+
+
+@pytest.mark.asyncio
+async def test_set_viewport_persists_on_navigation(bidi_session, new_tab, inline):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:47]: Would you consider to parameterize this method as suggested so that we also have a coop navigation? Here the existing browsing context will definitely be replace which will not happen for a navigation like the one we have right now.

**Review by whimboo — CHANGES_REQUESTED** [review]
Thanks for the update. I think it's close. Just one question for a test enhancement and a couple of nits.

@OrKoN am I correct that you will create another PR for the invalid cases? Those should be pretty easy to add and having it all in a single PR might be nice.

**Comment by whimboo** [comment]
Please also check the currently failing lint and unit test jobs for other not mentioned issues.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 7 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:48]: yeah, I am still looking into that

**Comment by OrKoN** [comment]
Add invalid.py with a few invalid invocations for now

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import pytest
+
+import webdriver.bidi.error as error
+
+
+@pytest.mark.parametrize("value", [None, False, 42, {}, []])
+@pytest.mark.asyncio
+async def test_params_context_invalid_context(bidi_session, value):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:8]: Can you please use the same naming convention as what we have for all the other BiDi API's?
```suggestion
async def test_params_context_invalid_type(bidi_session, value):
```

We also need a test for an invalid value.

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    with pytest.raises(error.InvalidArgumentException):
+        await bidi_session.browsing_context.set_viewport(context=value, viewport={
+            "width": 100,
+            "height": 200,
+        })
+
+
+@pytest.mark.parametrize("viewport", [{"width": -1, "height": -1}, {"width": None, "height": 300}, {"width": 1200, "height": "String"}])
+@pytest.mark.asyncio
+async def test_params_context_invalid_viewport(bidi_session, new_tab, viewport):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:18]: Mind differentiating between invalid type and invalid value? Also we have to test height and width in combination so that we know that also height alone set to `None` will fail. Other invalid tests of existing APIs should give a good sense how we did that.

The test name should then start with `test_params_viewport_invalid_` as well.

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    with pytest.raises(error.InvalidArgumentException):
+        await bidi_session.browsing_context.set_viewport(context=new_tab["context"], viewport=viewport)
+
+
+@pytest.mark.asyncio
+async def test_invalid_frame(bidi_session):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:24]: To be in sync with other tests lets rename and move up.

```suggestion
async def test_params_context_invalid_value(bidi_session):
```

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    with pytest.raises(error.NoSuchFrameException):
+        await bidi_session.browsing_context.set_viewport(context="_invalid_")
+
+
+@pytest.mark.asyncio
+async def test_closed_frame(bidi_session, top_context, inline, add_and_remove_iframe):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:30]: This test was added only once so far for screenshots. No other API has it yet. I would say that we add it later on for all the other APIs in on go? 

Instead I would propose a test that creates a frame but doesn't remove it. Calling `setViewport` for the frame has to fail because we only allow top-level browsing contexts to be modified.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+      context=new_tab["context"],
+      viewport=None)
+    assert await get_viewport_dimensions(bidi_session, new_tab) == original_viewport
+
+
+@pytest.mark.asyncio
+async def test_set_viewport_affects_specific_context(bidi_session, top_context, new_tab):
+    original_viewport = await get_viewport_dimensions(bidi_session, top_context)
+
+    test_viewport = {"width": 333, "height": 333}
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:38]: You updated the other case but here the height and width are still the same.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    test_viewport = {"width": 499, "height": 599}
+
+    await bidi_session.browsing_context.set_viewport(
+        context=new_tab["context"],
+        viewport=test_viewport)
+
+    assert await get_viewport_dimensions(bidi_session, new_tab) == test_viewport
+
+    url = inline("<div>foo</div>", parameters=parameters, protocol=protocol, domain="alt")
+    await bidi_session.browsing_context.navigate(
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:67]: We actually need two navigations to test this scenario. First navigate to the page without using `domain`, and setup the viewport. Then navigate to the alternative domain to trigger a COOP navigation. Right now we navigate from `about:blank` which might not trigger it.

**Review by whimboo — CHANGES_REQUESTED** [review]
Great updates Alex! I think the PR is pretty close but lets make sure we stay in sync regarding the naming convention for other tests. That makes it easier to do mass-changes in the future.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 13 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:67]: Updated

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py (hunk 11 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:30]: I will remove the test and do a new test separately. The fixture for adding/removing frames needs refactoring as it is combined in one for some reason.

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py (hunk 9 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:18]: Python's type checking didn't allow me to actually test things like missing height or missing width. Should I rewrite the set_viewport client implementation to now expect specific keys from the test?

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 12 above)
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:38]: My comment is not about the new tab. The point is if due to some reason the remote end takes eg the value for width and applies it to both width and height we won't notice that if the value is equal.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py [hunk 14]
```diff
@@ -35,3 +36,25 @@ async def test_capture(bidi_session, top_context, inline, compare_png_bidi):
 
     comparison = await compare_png_bidi(new_data, data)
     assert comparison.equal()
+
+@pytest.mark.asyncio
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py:40]:
```suggestion

@pytest.mark.asyncio
```

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py [hunk 15]
```diff
@@ -35,3 +36,25 @@ async def test_capture(bidi_session, top_context, inline, compare_png_bidi):
...
+async def test_capture_with_viewport(bidi_session, new_tab, inline, compare_png_bidi):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py:41]: Probably we should parameterize this test to have a smaller and a larger size for the viewport compared to the original values.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py [hunk 16]
```diff
@@ -35,3 +36,25 @@ async def test_capture(bidi_session, top_context, inline, compare_png_bidi):
...
+    dpr = await get_device_pixel_ratio(bidi_session, new_tab)
+
+    test_viewport = {"width": 499, "height": 599}
+    await bidi_session.browsing_context.set_viewport(
+        context=new_tab["context"],
+        viewport=test_viewport)
+
+    expected_size = {
+        "width": test_viewport["width"] * dpr,
+        "height": test_viewport["height"] * dpr
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py:51]: Please use `floor` to cut of the fragments from the result given that dpr is a float value.

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+
+@pytest.mark.asyncio
+async def test_params_context_invalid_value(bidi_session):
+    with pytest.raises(error.NoSuchFrameException):
+        await bidi_session.browsing_context.set_viewport(context="_invalid_")
+
+
+@pytest.mark.parametrize("viewport", [{"width": None, "height": 300}, {"width": 1200, "height": None}, {"width": 1200, "height": "String"}])
+@pytest.mark.asyncio
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:23]: There are different checks that we have to perform.

1. types other than `{}` for the viewport itself.
2. an empty object, or only height or width
3. a loop for all the invalid values for height and width checked separately while giving a valid value for the other field.

With all that we won't need a parameterization of the test.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 12 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:38]: This is covered by the other test (test_set_viewport), right?

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+async def test_params_viewport_invalid_type(bidi_session, new_tab, viewport):
+    with pytest.raises(error.InvalidArgumentException):
+        await bidi_session.browsing_context.set_viewport(context=new_tab["context"], viewport=viewport)
+
+
+@pytest.mark.parametrize("viewport", [{"width": -1, "height": -1}])
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:29]: This is just a single test. As such there is no need to parameterize.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 12 above)
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:38]: But can you guarantee that there is no special situation when this could happen? Most likely not. It doesn't hurt to say eg. `499, 599` as for other tests.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 13 above)
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:67]: It seems that it hasn't pushed yet?

**Review by whimboo — CHANGES_REQUESTED** [review]

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 13 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:74]: Should be pushed now

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py (hunk 16 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py:51]: dpr should not be a float (see get_device_pixel_ratio helper where it is floored)

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py (hunk 17 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:23]: See my comment https://github.com/web-platform-tests/wpt/pull/40584#discussion_r1238613000 (Python does not allow sending those cases given how I implemented the client)

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py (hunk 9 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/invalid.py]: relaxed the client implementation and added more cases

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py (hunk 17 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/invalid.py]: Added those cases.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py (hunk 15 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py]: Added

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py (hunk 16 above)
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py:59]: So that was most likely an oversight and we shouldn't have floored it there. In case of a DPR of `1.5` and a viewport width of 100 we current would get `100` pixels, while it should be `150` pixels. Maybe we should better improve `get_viewport_dimensions` by adding a eg. `useDevicePixelRatio` flag to it which directly computes it right. What do you think? IMHO it should be fine in a follow-up.

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 19]
```diff
@@ -127,3 +127,14 @@ def print(self,
...
         return result["data"]
+
+    @command
+    def set_viewport(self, context: str, viewport: Optional[Mapping[str, Any]] = None) -> Mapping[str, Any]:
+        params: MutableMapping[str, Any] = {"context": context}
+
+        if viewport is not None:
+            params["viewport"] = viewport
+        else:
+            params["viewport"] = None
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:138]: Sorry to mention it now but we should make that similar to the `SerializationOptions` arguments and class that is used in the `script` BiDi module.

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+        await bidi_session.browsing_context.set_viewport(context=new_tab["context"], viewport={
+            "width": 100,
+            "height": height,
+        })
+
+
+@pytest.mark.asyncio
+async def test_params_viewport_invalid_value(bidi_session, new_tab):
+    with pytest.raises(error.InvalidArgumentException):
+        await bidi_session.browsing_context.set_viewport(context=new_tab["context"], viewport={"width": -1, "height": -1})
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:52]: nit: as best it should be tested separately to ensure that the boundary check is run for both individually.

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/invalid.py]: Updated

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py (hunk 19 above)
Comment (OrKoN) [reply tools/webdriver/webdriver/bidi/modules/browsing_context.py]: What does the use of the class like that enables?

Comment (whimboo) [reply tools/webdriver/webdriver/bidi/modules/browsing_context.py]: By using that class we could have tests that set only individual fields, but not accepting any kind of dict as in the current implementation. Means we exclude any extensible. Being in sync here with the script module would be good.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py (hunk 16 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py]: Updated in this PR. I wonder if browser implementations would floor or ceil when taking a screenshot 🤔

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 21]
```diff
@@ -127,3 +127,7 @@ def print(self,
     def _print(self, result: Mapping[str, Any]) -> Any:
         assert result["data"] is not None
         return result["data"]
+
+    @command
+    def set_viewport(self, context: str, viewport: Optional[Mapping[str, Any]] = None) -> Mapping[str, Any]:
+        return {"context": context, "viewport": viewport}
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:133]: Did you consider against using a `Viewport` class given that it is not used here, or haven't you updated that part yet?

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+async def test_params_viewport_height_invalid_type(bidi_session, new_tab, height):
+    with pytest.raises(error.InvalidArgumentException):
+        await bidi_session.browsing_context.set_viewport(context=new_tab["context"], viewport={
+            "width": 100,
+            "height": height,
+        })
+
+
+@pytest.mark.parametrize("viewport", [{"width": -1, "height": -1}, {"width": -1, "height": 100}, {"width": 100, "height": -1}])
+@pytest.mark.asyncio
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:50]: We should use ids to get properly constructed test names:

```suggestion
@pytest.mark.parametrize("viewport", [
    {"width": -1, "height": 100}, 
    {"width": 100, "height": -1},
    {"width": -1, "height": -1},     
], ids=["width negative", "height negative", "both negative"])
@pytest.mark.asyncio
```

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py (hunk 21 above)
Comment (OrKoN) [reply tools/webdriver/webdriver/bidi/modules/browsing_context.py:133]: I am still trying to understand the benefits of using a class: since we want to test that the implementation reacts properly for any input from the client, it seems that using a class similar to SerializationOptions would restrict that as it does not allow sending None values, for example. So it looks like the current version allows testing any input. Or have I misunderstood how the proposed class would be used?

Comment (whimboo) [reply tools/webdriver/webdriver/bidi/modules/browsing_context.py:133]: Oh, good point. So yeah, looks like that this is indeed the most flexible approach and that we may have to update the script module as well to offer more flexibility for the `SerializationOptions`. As such lets keep what we have here now.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py [hunk 23]
```diff
@@ -0,0 +1,107 @@
+import pytest
+
+from ... import get_viewport_dimensions
+
+
+@pytest.mark.asyncio
+async def test_set_viewport(bidi_session, new_tab):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:7]: If you don't mind maybe you can create one further test which sets the viewport to the exact same dimensions? I noticed that I had a bug in our Firefox implementation which caused the command to hang because it was waiting for a resize of the viewport.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py [hunk 24]
```diff
@@ -35,3 +37,33 @@ async def test_capture(bidi_session, top_context, inline, compare_png_bidi):
 
     comparison = await compare_png_bidi(new_data, data)
     assert comparison.equal()
+
+
+@pytest.mark.parametrize("delta_width", [-10, +20])
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py:42]: Actually would you mind to move this test to it's own file eg `viewport.py` in this folder? That way I could also already prepare easily a metadata file for us.

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py (hunk 23 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:7]: Added

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py (hunk 24 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py]: Added

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py [hunk 25]
```diff
@@ -39,8 +39,8 @@ async def test_capture(bidi_session, top_context, inline, compare_png_bidi):
     assert comparison.equal()
 
 
-@pytest.mark.parametrize("delta_width", [-10, +20])
-@pytest.mark.parametrize("delta_height", [-30, +40])
+@pytest.mark.parametrize("delta_width", [-10, +20], ids=["width smaller than original", "width larger than original"])
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py:42]: Note that this makes the testname quite long to something like: `test_capture_with_viewport[width smaller than original-height smaller than original", "height larger than original].

Maybe we can reduce to `ids=["width smaller", ...]`? I think that should be fine.

**Review by whimboo — APPROVED** [review]
Looks great for me now. Especially having about 40 new tests for this specific command!

Just two nits. Feel free to apply or not. Thanks a lot for working on this PR!

File: webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py (hunk 25 above)
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/capture_screenshot/capture_screenshot.py:42]: Done

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py [hunk 26]
```diff
@@ -16,6 +16,25 @@ async def test_set_viewport(bidi_session, new_tab):
     assert await get_viewport_dimensions(bidi_session, new_tab) == test_viewport
 
 
+@pytest.mark.asyncio
+async def test_set_viewport_twice(bidi_session, new_tab):
```
[the remark this answers is not in the document]
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py]: Done

**Comment by whimboo** [comment]
@OrKoN can you get someone from Google to review as well?

File: webdriver/tests/bidi/browsing_context/set_viewport/invalid.py
Comment (thiagowfx) [inline webdriver/tests/bidi/browsing_context/set_viewport/invalid.py:1]: LGTM

**Review by thiagowfx — APPROVED** [review]

File: webdriver/tests/bidi/__init__.py [hunk 27]
```diff
@@ -86,11 +86,11 @@ async def get_device_pixel_ratio(bidi_session, context):
 
     :param bidi_session: BiDiSession
     :param context: Browsing context ID
-    :returns: (int) devicePixelRatio.
+    :returns: (float) devicePixelRatio.
```
[the remark this answers is not in the document]
Comment (OrKoN) [reply webdriver/tests/bidi/__init__.py:89]: I don't want to change it in this PR as I am not certain that it would work everywhere. Do you want to open a follow-up? I bet there are other places where a similar change can be applied?

[the remark this answers is not in the document]
Comment (thiagowfx) [reply webdriver/tests/bidi/__init__.py:89]: OK sure. I'll add it to my ongoing PR that deals with similar changes then.

**Comment by whimboo** [comment]
@jgraham could you please admin merge this PR? Thanks.
